src/Tui/ZTextLayout.cpp

```diff
@@ -629,18 +636,69 @@ void ZTextLineRef::draw(ZPainter painter, const QPoint &pos, ZTextStyle color, Z
+        for (int i = 0; i < ld.textRuns.size(); i++) {
+            textRunStarts.push_back(ld.textRuns[i].offset);
+        }
+
+        // match each format range to all text runs that intersect it.
+        // This can add multiple entries per orignial format range.
+        std::vector<HighlightSortHelper> partitionedRanges;
+        partitionedRanges.reserve(ranges.size()); // enough if no format range intersects multiple text runs
+
+        for (int i = 0; i < ranges.size(); i++) {
+            const ZFormatRange &formatRange = ranges[i];
+            const auto formatRangeLength = formatRange.length();
+            if (formatRangeLength <= 0) {
+                continue;
+            }
+            auto formatRangeStart = formatRange.start();
+            int formatRangeEnd = formatRangeStart + formatRangeLength;
+
+            if (formatRangeEnd < 0 || formatRangeStart >= ld.endIndex) {
+                continue;
+            }
+
+            if (formatRangeStart < ld.offset) {
+                formatRangeStart = ld.offset;
+            }
+
+            auto lb = std::upper_bound(textRunStarts.begin(), textRunStarts.end(), formatRangeStart);
+            lb--; // safe as frStart >= ld.offset and *textRunStarts.begin() == ld.offset
+            int textRunIndex = lb - textRunStarts.begin();
+            for (; textRunIndex < ld.textRuns.size(); textRunIndex++) {
+                partitionedRanges.push_back(HighlightSortHelper{textRunIndex, &formatRange});
+                if (ld.textRuns[textRunIndex].endIndex >= formatRangeEnd) {
+                    break;
+                }
+            }
+        }
+
+        // Sort by run index, so that the rendering code can iterate over the format ranges in O(1)
+        std::stable_sort(partitionedRanges.begin(), partitionedRanges.end(),
+                         [] (const HighlightSortHelper &lhs, const HighlightSortHelper &rhs) {
+            return lhs.run < rhs.run;
+        });
+
+
+        auto nextFormatRange = partitionedRanges.begin();
         for (int i = 0; i < ld.textRuns.size(); i++) {
             const ZTextLayoutPrivate::TextRun &run = ld.textRuns[i];
             if (run.type == ZTextLayoutPrivate::TextRun::COPY) {
                 painterClipped.writeWithAttributes(pos.x() + ld.pos.x() + run.x, pos.y() + ld.pos.y(),
                                                    p->text.mid(run.offset, run.endIndex - run.offset),
                                                    color.foregroundColor(), color.backgroundColor(), color.attributes());
-                for (const ZFormatRange &formatRange: ranges) {
-                    if (formatRange.length() <= 0) {
+                for (; nextFormatRange != partitionedRanges.end() && nextFormatRange->run == i; nextFormatRange++) {
+                    const ZFormatRange &formatRange = *nextFormatRange->ptr;
+                    const auto formatRangeLength = formatRange.length();
+                    if (formatRangeLength <= 0) {
```

Comment:
is this condition (and similar conditions below) even necessary? `partitionedRanges?  should not contain ranges with length <= 0: https://github.com/tuiwidgets/tuiwidgets/pull/5/files#diff-4339a5b85f937a0c5042d123203183c05fa37800d612fcc9a0b00ce78bf34dbbR654

Reply:
Right, that was a leftover that is no longer needed.